igvm/vm.py

```diff
@@ -849,3 +866,94 @@ def vm_performance_value(self) -> float:
+
+        :param: timeout: Timeout value for the head/get request
+
+        :return: VM types overview as list
+        """
+
+        url = AWS_INSTANCES_OVERVIEW_URL
+        file = Path(AWS_INSTANCES_OVERVIEW_FILE)
+        etag_file = Path(AWS_INSTANCES_OVERVIEW_FILE_etag)
+
+        head_req = Request(url, method='HEAD')
+        resp = urlopen(head_req, timeout=timeout)
+        if resp.status == 200:
+            etag = dict(resp.info())['ETag']
+        else:
+            log.debug('Could not retrieve ETag from {}'.format(url))
+
+        if file.exists():
+            if etag_file.exists():
+                with open(AWS_INSTANCES_OVERVIEW_FILE_etag) as f:
+                    prev_etag = f.read()
+                if etag and etag == prev_etag:
+                    with open(AWS_INSTANCES_OVERVIEW_FILE) as f:
+                        return json.load(f)
+
+        resp = urlopen(url, timeout=timeout)
+        if etag:
+            with open(AWS_INSTANCES_OVERVIEW_FILE_etag, 'w') as f:
+                f.write(etag)
+        with open(AWS_INSTANCES_OVERVIEW_FILE, 'w') as f:
+            f.write(resp.read().decode('utf-8'))
+        with open(AWS_INSTANCES_OVERVIEW_FILE) as f:
+            return json.load(f)
+
+
+    def aws_get_fitting_vm_types(self, overview: List) -> List:
+        """AWS Get Fitting VM types
+
+        Use the performance_value of the VM and multiply it with a static
+        factor to get the targeted ECU of AWS +-25% (which is kind of a
+        performance_value). Crawl the instances overview for fitting
+        vm_types for this targeted ECU range.
+
+        :param: overview: instances.json with a complete
+                          instance_types overview
+
+        :return: Fitting VM types as list
+        """
+
+        vm_performance_value = self.vm_performance_value()
+        region = self.dataset_obj['aws_placement'][:-1]
+
+        ecu_targeted = {
+            'min':
+                (vm_performance_value * AWS_ECU_FACTOR) -
+                (vm_performance_value * AWS_ECU_FACTOR * 0.25),
+            'max':
+                (vm_performance_value * AWS_ECU_FACTOR) +
+                (vm_performance_value * AWS_ECU_FACTOR * 0.25)
+        }
+
+        vm_types = dict()
+
+        for t in overview:
+            vm_types[t['instance_type']] = dict()
+            vm_types[t['instance_type']]['instance_type'] = t['instance_type']
+            if t['memory'] >= (self.dataset_obj['memory'] / 1024):
+                vm_types[t['instance_type']]['memory'] = t['memory']
+            if 'ECU' in t and isinstance(t['ECU'], (int, float)):
+                if ecu_targeted['min'] <= t['ECU'] <= ecu_targeted['max']:
+                    vm_types[t['instance_type']]['ECU'] = t['ECU']
+            if region in t['pricing']:
+                vm_types[t['instance_type']]['price'] = float(
+                    t['pricing'][region]['linux']['ondemand'])
+
+            if len(vm_types[t['instance_type']]) < 4:
```

Comment:
sorry i did not catch this earlier. i would rewrite this loop a little bit because it is kinda unclear at first what this condition is really for. we can also prevent unnecessarily adding and again removing entries from the main dict.
```
for t in overview:
    if region not in t['pricing']:
        continue
    if t['memory'] < (self.dataset_obj['memory'] / 1024):
        continue
    if 'ECU' not in t or not isinstance(t['ECU'], (int, float)):
        continue

    # now it gets very straightforward:
    vm_types[t['instance_type']] = dict(
        instance_type=t['instance_type'],
        memory=t['memory'],
        ECU=t['ECU'],
        price=float(t['pricing'][region]['linux']['ondemand']),
    )
```

Reply:
actually the last statement could be as simple as `vm_types[t['instance_type']] = t` if you like

Reply:
That's actual even nicer. Thanks for the cleanup:)